packages/react/checkbox/src/Checkbox.tsx

```diff
@@ -130,28 +134,24 @@ type CheckboxIndicatorOwnProps = {
   forceMount?: true;
 };
 
-const CheckboxIndicator = forwardRefWithAs<typeof CheckboxIndicatorImpl, CheckboxIndicatorOwnProps>(
-  (props, forwardedRef) => {
-    const { forceMount, ...indicatorProps } = props;
-    const checked = useCheckboxContext(INDICATOR_NAME);
-    return (
-      <Presence present={forceMount || checked === 'indeterminate' || checked === true}>
-        <CheckboxIndicatorImpl
-          {...indicatorProps}
-          data-state={getState(checked)}
-          ref={forwardedRef}
-        />
-      </Presence>
-    );
-  }
-);
-
-const CheckboxIndicatorImpl = forwardRefWithAs<typeof INDICATOR_DEFAULT_TAG>(
-  function CheckboxIndicatorImpl(props, forwardedRef) {
-    const { as: Comp = INDICATOR_DEFAULT_TAG, ...indicatorProps } = props;
-    return <Comp {...indicatorProps} {...getPartDataAttrObj(INDICATOR_NAME)} ref={forwardedRef} />;
-  }
-);
+const CheckboxIndicator = forwardRefWithAs<
+  typeof INDICATOR_DEFAULT_TAG,
+  MergeOwnProps<typeof Primitive, CheckboxIndicatorOwnProps>
+>((props, forwardedRef) => {
+  const { forceMount, ...indicatorProps } = props;
+  const checked = useCheckboxContext(INDICATOR_NAME);
+  return (
+    <Presence present={forceMount || checked === 'indeterminate' || checked === true}>
+      <Primitive
```

Comment:
Yeah, that was a neat idea but tbh, the abstraction here actually confused me when I was doing this because it made me think something more complex was going on. I spent some time trying to figure out what that was.

If there is no need for that additional mental computation I think it's helpful to just avoid it by keeping it inline till we know the abstraction is needed.

We just need to make sure we keep on top of that in PR reviews I guess 😬

Reply:
Yeah I was thinking the same. I feel like it's just something that is pretty easy to catch or eventually realise that we're doing too much work in a component that's always rendered, and in this case remediate. Rather than premeditate the changes.

Reply:
💯 :+1: